### DISCUSSION TITLE
How to get all bindings?

### DISCUSSION THREAD

**Discussion by bunder** [justinethier/cyclone#523] [General]
Is it possible to get list of all current bindings? I want something like this:

```
(define asdf 4)
(display (env:all-variables <something>))
```

And expect to see `asdf` there.

**Comment by justinethier** [comment]
How about this:

```
cyclone> (define a 1)
ok
cyclone> (define b 2)
ok
cyclone> (define c 3)
ok
cyclone> (env:all-variables *global-environment*)
(c b a)
```

**Reply by bunder** [reply]
Thank you! This works in an interpreter. How to make it work in a compiled program? Now I get following error:

```
Error: Unbound variable(s):
(*global-environment* env:all-variables)
```

**Reply by justinethier** [reply]
Well, what are you trying to do?

**Reply by bunder** [reply]
Finally I want to define some number of functions with common prefix, for example `foo-bar`, `foo-baz` and `foo-qux`. And then I want generic code that will call all these functions and print their results.

**Reply by justinethier** [reply]
Couple of ideas:

- You could keep track of functions as you define them, such as in this library: https://github.com/justinethier/cyclone/blob/master/libs/cyclone/foreign.sld#L51
- Alternatively, could you generate a library and import everything from that library? I am thinking this may not quite be what you are looking for, though.

Thoughts?

**Reply by bunder** [reply]
Thank you! Yes, I could always wrap defines and add all functions to a specific list, but this will not work for functions that are defined outside of my code using standard `define`. Anyway, despite this is non-standard to R7RS, my goal is to understand whether getting all symbols is possible in compiled programs in Cyclone?

**Reply by justinethier** [reply]
Are you looking to get the symbols at compile time or runtime?

**Reply by lassik** [reply]
Poking around in the symbol namespace is generally not a sound way to solve problems in Lisp languages. Scheme doesn't have a clear idea of what "the namespace" is. For production code, the namespaces of interest are libraries, and things outside libraries (e.g. in the REPL environment) tend to be experimental stuff that production code should probably not touch unless told to do so.

**Reply by bunder** [reply]
> Are you looking to get the symbols at compile time or runtime?

I just experiment with Cyclone and looking for any possibility. As I understand, Scheme doesn't specify standard way to get all bindings, but implementations may provide it on they own (like `(current-module)` in Guile).

**Reply by bunder** [reply]
> Poking around in the symbol namespace is generally not a sound way to solve problems in Lisp languages. ...

Generally you are right, but I just want to understand if this feature is supported in Cyclone.
